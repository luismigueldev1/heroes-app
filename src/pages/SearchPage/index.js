import React, { useMemo } from "react";
import queryString from "query-string";
import HeroesListItem from "../../components/HeroesListItem";

import { useForm } from "../../hooks/useForm";
import { getHeroesByName } from "../../selectors/getHeroesByName";

export default function SearchPage({ history, location }) {
  const { q = "" } = queryString.parse(location.search);

  const [values, handleInputChange] = useForm({
    search: q,
  });

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  const handleSubmit = (event) => {
    event.preventDefault();
    history.push(`?q=${values.search}`);
  };

  return (
    <div>
      <h1>Search Heroes</h1>
      <hr />

      <div className="row">
        <div className="col-sm-12 col-md-5">
          <form onSubmit={handleSubmit}>
            <label htmlFor="search" className="form-label">
              Find a superhero here:
            </label>
            <input
              type="text"
              name="search"
              className="form-control"
              value={values.search}
              onChange={handleInputChange}
            />
            <div className="d-grid">
              <button type="submit" className="btn mt-1 btn btn-primary">
                Search
              </button>
            </div>
          </form>
        </div>
        <div className="col-sm-12 col-md-7">
          <div className="row">
            {q !== "" && heroesFiltered.length === 0 && (
              <div className="alert alert-warning" role="alert">
                There's not a superhero by the name: {q}
              </div>
            )}

            {q === "" && (
              <div className="alert alert-info" role="alert">
                Search some hero first
              </div>
            )}
            {heroesFiltered.map((hero) => (
              <HeroesListItem hero={hero} key={hero.id} />
            ))}
          </div>
        </div>
      </div>
    </div>
  );
}
