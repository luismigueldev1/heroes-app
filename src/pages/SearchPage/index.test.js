import React from "react";
import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import SearchPage from ".";

describe("Pruebas en <SearchPage/>", () => {
  const query = "batman121212";

  test("debe de mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchPage} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toBe(
      "Search some hero first"
    );
  });

  test("debe de mostrar a batman y el input con el valor del queryString", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={SearchPage} />
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
  });

  test("debe de mostrar un error si no se encuetra el hero", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${query}`]}>
        <Route path="/search" component={SearchPage} />
      </MemoryRouter>
    );

    expect(wrapper.find(".alert-warning").text().trim()).toBe(
      `There's not a superhero by the name: ${query}`
    );
  });

  test("debe de llamar el push del history", () => {
    const historyMock = {
      push: jest.fn(),
      location: {},
      listen: jest.fn(),
      createHref: jest.fn(),
      replace: jest.fn(),
    };

    const locationMock = {};

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${query}`]}>
        <Route
          path="/search"
          component={() => (
            <SearchPage history={historyMock} location={locationMock} />
          )}
        />
      </MemoryRouter>
    );

    wrapper.find("input").simulate("change", {
      target: {
        name: "search",
        value: "batman",
      },
    });

    wrapper.find("form").simulate("submit", { preventDefault: () => {} });

    expect(historyMock.push).toHaveBeenCalledWith(`?q=batman`);
  });
});
