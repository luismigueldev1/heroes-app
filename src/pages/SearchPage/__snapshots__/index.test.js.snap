// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage/> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Heroes
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-sm-12 col-md-5"
    >
      <form
        onSubmit={[Function]}
      >
        <label
          className="form-label"
          htmlFor="search"
        >
          Find a superhero here:
        </label>
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          type="text"
          value=""
        />
        <div
          className="d-grid"
        >
          <button
            className="btn mt-1 btn btn-primary"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-sm-12 col-md-7"
    >
      <div
        className="row"
      >
        <div
          className="alert alert-info"
          role="alert"
        >
          Search some hero first
        </div>
      </div>
    </div>
  </div>
</div>
`;
